### PR TITLE
Report junctions

### PR DESCRIPTION
**Pull request by afshinfard**
aim: report junction-causing barcodes back to molecule stage so this stage do another round of molecule separation (more extensive this time) on only those vertices.
- a new method to report the junctions
- `determine_molecules()` now handles a second input file for junctions (as a graph with no edge) and then applies the requested set of algorithms only on those barcodes. It applies `bc` on all other barcodes not in junctions.
- and unrelated commit merged here but with no effect: `split_junctions_of_tree()` has a new input argument. when set to 0 it removes the junction instead of keeping the two largest edges. The default is set to 1 so no effect.
- I will add `Makefile` codes for reporting junctions and redoing mol-separation later.

### CODE REVIEW

**Comment by afshinfard** [comment]
@lcoombe It's not what @jowong4 asked for and that one is already pushed into master. But I appreciate if you can review it if you had time.

File: physlr/physlr.py [hunk 1]
```diff
@@ -1276,6 +1287,25 @@ def physlr_mst_troubleshoot(self):
...
+        gmst = Physlr.determine_pruned_mst(g)
+        print(int(timeit.default_timer() - t0), "Searching for junctions...", file=sys.stderr)
+        junctions = []
+        for component in nx.connected_components(gmst):
+            junctions += Physlr.detect_junctions_of_tree(
+                gmst.subgraph(component), Physlr.args.prune_junctions)
+        print(int(timeit.default_timer() - t0),
+              "Found", len(junctions), "junctions.", file=sys.stderr)
+        gjunctions = nx.Graph()
+        gjunctions.add_nodes_from(junctions)
```
Comment (lcoombe) [inline physlr/physlr.py]: Just wondering -- is it necessary to format this output as a graph? It looks like you are just adding nodes, but no edges, so couldn't you just use a TSV file of the junction nodes?

Comment (afshinfard) [reply physlr/physlr.py]: Right, I could. I found it easier at that time to do it that way as the method `write_tsv()` writes graphs in tsv format. I'll add a few lines of code to write the list as tsv.

File: physlr/physlr.py [hunk 2]
```diff
@@ -537,9 +537,13 @@ def split_junctions_of_tree(prune_junctions, gin):
             if Physlr.args.verbose >= 3:
                 print("Junction:", junction, "Edges:", *edges, file=sys.stderr)
             # Keep the two incident edges with the largest weight, and remove the rest.
-            for u, v, _ in edges[2:-1]:
-                g.remove_edge(u, v)
-        return g
+            if keep_largest:
+                for u, v, _ in edges[2:-1]:
```
Comment (jowong4) [inline physlr/physlr.py]: shouldn't this code be
```suggestion
                for u, v, _ in edges[1:]:
```

File: physlr/physlr.py [hunk 3]
```diff
@@ -1276,6 +1287,23 @@ def physlr_mst_troubleshoot(self):
...
+        gmst = Physlr.determine_pruned_mst(g)
+        print(int(timeit.default_timer() - t0), "Searching for junctions...", file=sys.stderr)
+        junctions = []
+        for component in nx.connected_components(gmst):
+            junctions += Physlr.detect_junctions_of_tree(
+                gmst.subgraph(component), Physlr.args.prune_junctions)
+        print(int(timeit.default_timer() - t0),
+              "Found", len(junctions), "junctions.", file=sys.stderr)
+        for junction in junctions:
+            print(junction, file=sys.stdout)
```
Comment (jowong4) [inline physlr/physlr.py]:
```suggestion
            print(junction)
```
I don't think the `file=sys.stdout` is necessary

File: physlr/physlr.py [hunk 4]
```diff
@@ -1276,6 +1287,23 @@ def physlr_mst_troubleshoot(self):
...
+        print(int(timeit.default_timer() - t0), "Wrote the junctions.", file=sys.stderr)
```
Comment (jowong4) [inline physlr/physlr.py]:
```suggestion
        print(int(timeit.default_timer() - t0), "Wrote junctions to file.", file=sys.stderr)
```

File: physlr/physlr.py [hunk 5]
```diff
@@ -1689,21 +1720,36 @@ def physlr_molecules(self):
...
                       "--separation-strategy: " + str(set(alg_list) - alg_white_list)
             sys.exit(exit_message)
+        junctions = []
+        if len(self.args.FILES) > 1:
+            with open(self.args.FILES[1]) as fin:
+                for line in fin:
+                    junctions.append(line.split()[0])
+            print(
+                int(timeit.default_timer() - t0),
+                "Separating junction resulting barcodes into molecules"
```
Comment (jowong4) [inline physlr/physlr.py]:
```suggestion
                "Separating junction (contributing / causing) barcodes into molecules"
```
(contributing / causing)  I think sounds better than resulting

File: physlr/physlr.py [hunk 6]
```diff
@@ -1689,21 +1720,36 @@ def physlr_molecules(self):
             exit_message = "Error: physlr molecule: wrong input parameter(s) " + \
                       "--separation-strategy: " + str(set(alg_list) - alg_white_list)
             sys.exit(exit_message)
+        junctions = []
+        if len(self.args.FILES) > 1:
+            with open(self.args.FILES[1]) as fin:
```
[the remark this answers is not in the document]
Comment (jowong4) [reply physlr/physlr.py]: Yeah I think there is a bug here, gin doesn't get initialized if the if statement is true

[the remark this answers is not in the document]
Comment (afshinfard) [reply physlr/physlr.py]: Oh right, when I changed the code I forgot to consider this.

File: physlr/physlr.py (hunk 2 above)
Comment (afshinfard) [reply physlr/physlr.py]: It keeps removes all the edges except the first two edges. so skipping 0 and 1 it should iterate from 2 to last. so [2:]. I'll fix it.

File: physlr/physlr.py (hunk 3 above)
Comment (afshinfard) [reply physlr/physlr.py]: right

File: physlr/physlr.py (hunk 2 above)
Comment (jowong4) [reply physlr/physlr.py]: It said `keep_largest` so I thought you only wanted the largest edge hence remove everything from 1 to last. So do you want to keep the two largest?

File: physlr/physlr.py (hunk 5 above)
Comment (afshinfard) [reply physlr/physlr.py]: should I join them with a dash? like junction-causing?

Comment (lcoombe) [reply physlr/physlr.py]: Yeah I'd prefer `junction-causing`

File: physlr/physlr.py (hunk 3 above)
Comment (lcoombe) [reply physlr/physlr.py]: We do specify `file=sys.stdout` throughout the physlr file though? I'd be OK keeping it for consistency.

File: physlr/physlr.py (hunk 2 above)
Comment (afshinfard) [reply physlr/physlr.py]: the largest link between two of the branches, that includes two edges. yep.

File: physlr/physlr.py (hunk 3 above)
Comment (afshinfard) [reply physlr/physlr.py]: Most of the time yes we do specif, and sometimes we don't. I'll keep it then.

**Comment by afshinfard** [comment]
The commit msg is not corrected! Don't consider it. Is there an easy way to change it?

File: physlr/physlr.py (hunk 2 above)
Comment (sjackman) [reply physlr/physlr.py]: Yes, that's a bug. It should be `[2:]`. What's the effect of that bug?

Comment (sjackman) [reply physlr/physlr.py]: Also 😠 😭

Comment (jowong4) [reply physlr/physlr.py]: hmm looks like I misunderstood the intention behind the code. That's all

**Comment by jowong4** [comment]
> The commit msg is not corrected! Don't consider it. Is there an easy way to change it?

You can rebase -i and change the commit messages. You will have push --force at the end though

File: physlr/physlr.py (hunk 2 above)
Comment (afshinfard) [reply physlr/physlr.py]: @sjackman no effects. This part of the code was not in the pipeline.

**Pull request merged**